File: main/src/main/java/com/google/android/apps/muzei/SourceSubscriberService.java
/*
 * Copyright 2014 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.apps.muzei;

import android.app.IntentService;
import android.app.PendingIntent;
import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;

import com.google.android.apps.muzei.api.MuzeiArtSource;
import com.google.android.apps.muzei.api.UserCommand;
import com.google.android.apps.muzei.api.internal.SourceState;
import com.google.android.apps.muzei.room.Artwork;
import com.google.android.apps.muzei.room.MuzeiDatabase;
import com.google.android.apps.muzei.room.Source;
import com.google.android.apps.muzei.sync.TaskQueueService;

import java.util.ArrayList;

import static com.google.android.apps.muzei.api.internal.ProtocolConstants.ACTION_PUBLISH_STATE;
import static com.google.android.apps.muzei.api.internal.ProtocolConstants.EXTRA_STATE;
import static com.google.android.apps.muzei.api.internal.ProtocolConstants.EXTRA_TOKEN;

public class SourceSubscriberService extends IntentService {
    private static final String TAG = "SourceSubscriberService";

    public SourceSubscriberService() {
        super("SourceSubscriberService");
    }

    @Override
    protected void onHandleIntent(Intent intent) {
        if (intent == null || intent.getAction() == null) {
            return;
        }

        String action = intent.getAction();
        if (!ACTION_PUBLISH_STATE.equals(action)) {
            return;
        }
        // Handle API call from source
        String token = intent.getStringExtra(EXTRA_TOKEN);
        Source source = MuzeiDatabase.getInstance(this).sourceDao().getCurrentSourceBlocking();
        if (source == null ||
                !TextUtils.equals(token, source.componentName.flattenToShortString())) {
            Log.w(TAG, "Dropping update from non-selected source, token=" + token
                    + " does not match token for " + source);
            return;
        }

        SourceState state = null;
        if (intent.hasExtra(EXTRA_STATE)) {
            Bundle bundle = intent.getBundleExtra(EXTRA_STATE);
            if (bundle != null) {
                state = SourceState.fromBundle(bundle);
            }
        }

        if (state == null) {
            // If there is no state, there is nothing to change
            return;
        }

        source.description = state.getDescription();
        source.wantsNetworkAvailable = state.getWantsNetworkAvailable();
        source.supportsNextArtwork = false;
        source.commands = new ArrayList<>();
        int numSourceActions = state.getNumUserCommands();
        for (int i = 0; i < numSourceActions; i++) {
            UserCommand command = state.getUserCommandAt(i);
            if (command.getId() == MuzeiArtSource.BUILTIN_COMMAND_ID_NEXT_ARTWORK) {
                source.supportsNextArtwork = true;
            } else {
                source.commands.add(command);
            }
        }

        com.google.android.apps.muzei.api.Artwork currentArtwork = state.getCurrentArtwork();
        if (currentArtwork != null) {
            MuzeiDatabase database = MuzeiDatabase.getInstance(this);
            database.beginTransaction();
            database.sourceDao().update(source);
            Artwork artwork = new Artwork();
            artwork.sourceComponentName = source.componentName;
            artwork.imageUri = currentArtwork.getImageUri();
            artwork.title = currentArtwork.getTitle();
            artwork.byline = currentArtwork.getByline();
            artwork.attribution = currentArtwork.getAttribution();
            artwork.token = currentArtwork.getToken();
            if (currentArtwork.getMetaFont() != null) {
                artwork.metaFont = currentArtwork.getMetaFont();
            }
            artwork.viewIntent = currentArtwork.getViewIntent();

            try {
                // Make sure we can construct a PendingIntent for the Intent
                PendingIntent.getActivity(this, 0, artwork.viewIntent,
                        PendingIntent.FLAG_UPDATE_CURRENT);
            } catch (RuntimeException e) {
                // This is actually meant to catch a FileUriExposedException, but you can't
                // have catch statements for exceptions that don't exist at your minSdkVersion
                Log.w(TAG, "Removing invalid View Intent that contains a file:// URI: " +
                        artwork.viewIntent, e);
                artwork.viewIntent = null;
            }

            database.artworkDao().insert(this, artwork);

            // Download the artwork contained from the newly published SourceState
            startService(TaskQueueService.getDownloadCurrentArtworkIntent(this));

            database.setTransactionSuccessful();
            database.endTransaction();
        }
    }
}

File: main/src/main/java/com/google/android/apps/muzei/sources/SourceSubscriberService.kt
/*
 * Copyright 2014 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.apps.muzei.sources

import android.app.IntentService
import android.content.Intent
import android.util.Log
import androidx.core.net.toUri
import androidx.room.withTransaction
import com.google.android.apps.muzei.api.MuzeiArtSource
import com.google.android.apps.muzei.api.internal.ProtocolConstants.ACTION_PUBLISH_STATE
import com.google.android.apps.muzei.api.internal.ProtocolConstants.EXTRA_STATE
import com.google.android.apps.muzei.api.internal.ProtocolConstants.EXTRA_TOKEN
import com.google.android.apps.muzei.api.internal.SourceState
import com.google.android.apps.muzei.api.provider.Artwork
import com.google.android.apps.muzei.api.provider.ProviderContract
import com.google.android.apps.muzei.room.MuzeiDatabase
import com.google.android.apps.muzei.room.SourceDao
import kotlinx.coroutines.runBlocking
import net.nurik.roman.muzei.BuildConfig
import net.nurik.roman.muzei.BuildConfig.SOURCES_AUTHORITY
import java.util.ArrayList

class SourceSubscriberService : IntentService("SourceSubscriberService") {

    companion object {
        private const val TAG = "SourceSubscriberService"
    }

    override fun onHandleIntent(intent: Intent?) {
        if (ACTION_PUBLISH_STATE != intent?.action) {
            return
        }
        // Handle API call from source
        val token = intent.getStringExtra(EXTRA_TOKEN)
        val state = intent.getBundleExtra(EXTRA_STATE)?.run {
            SourceState.fromBundle(this)
        } ?: return // If there's no state, there's nothing to change

        val database = MuzeiDatabase.getInstance(this)
        runBlocking {
            database.withTransaction {
                update(database.sourceDao(), token, state)
            }
        }
    }

    private suspend fun update(sourceDao: SourceDao, sourceToken: String, state: SourceState)  {
        val source = sourceDao.getCurrentSource()
        if (source == null || sourceToken != source.componentName.flattenToShortString()) {
            Log.w(TAG, "Dropping update from non-selected source, token=$sourceToken " +
                    "does not match token for ${source?.componentName}")
            return
        }

        source.apply {
            description = state.description
            wantsNetworkAvailable = state.wantsNetworkAvailable
            supportsNextArtwork = false
            commands = ArrayList()
        }
        val numSourceActions = state.numUserCommands
        for (i in 0 until numSourceActions) {
            val command = state.getUserCommandAt(i)
            when (state.getUserCommandAt(i)?.id) {
                MuzeiArtSource.BUILTIN_COMMAND_ID_NEXT_ARTWORK -> source.supportsNextArtwork = true
                else -> source.commands.add(command)
            }
        }
        sourceDao.update(source)

        val currentArtwork = state.currentArtwork
        if (currentArtwork?.imageUri != null) {
            val newArtwork = Artwork().apply {
                metadata = source.componentName.toShortString()
                persistentUri = currentArtwork.imageUri
                title = currentArtwork.title
                byline = currentArtwork.byline
                attribution = currentArtwork.attribution
                token = currentArtwork.imageUri.toString()
                webUri = currentArtwork.viewIntent?.toUri(Intent.URI_INTENT_SCHEME)?.toUri()
            }

            val artworkUri = ProviderContract.getProviderClient(this,
                    SOURCES_AUTHORITY).setArtwork(newArtwork)
            if (BuildConfig.DEBUG) {
                Log.d(TAG, "Set artwork to: $artworkUri")
            }
        } else if (currentArtwork != null) {
            Log.w(TAG, "Skipping artwork with an imageUri: $currentArtwork")
        }
    }
}
